File: diagflow/rag/embedder.py
"""
Embedding abstraction — converts text to vector embeddings.

Supports OpenAI embeddings (default) with a fallback to a simple
hash-based embedding for offline demo mode.
"""

from __future__ import annotations

import hashlib
import logging
import os
from typing import Any

import numpy as np

logger = logging.getLogger(__name__)


class Embedder:
    """Converts text to embedding vectors."""

    def __init__(
        self,
        model: str = "text-embedding-3-small",
        api_key: str | None = None,
        embedding_dim: int = 1536,
    ):
        self.model = model
        self.api_key = api_key or os.environ.get("OPENAI_API_KEY")
        # Fallback vectors must match the vector store's configured dimension
        # (embedding_dim in config), otherwise Milvus/Chroma upsert fails on
        # dimension mismatch when no embedding API key is present.
        self._dim = embedding_dim

    def embed(self, text: str) -> list[float]:
        """Embed text. Uses OpenAI if key available, otherwise fallback."""
        if self.api_key:
            return self._openai_embed(text)
        return self._fallback_embed(text)

    def embed_batch(self, texts: list[str]) -> list[list[float]]:
        return [self.embed(t) for t in texts]

    def _openai_embed(self, text: str) -> list[float]:
        try:
            from openai import OpenAI
            client = OpenAI(api_key=self.api_key)
            resp = client.embeddings.create(model=self.model, input=text)
            return resp.data[0].embedding
        except Exception:
            logger.warning("OpenAI embed failed, using fallback", exc_info=True)
            return self._fallback_embed(text)

    def _fallback_embed(self, text: str) -> list[float]:
        """Simple hash-based embedding for offline demo mode.

        Not semantically meaningful, but produces consistent vectors
        so identical/very similar texts cluster together.

        Emits ``self._dim`` elements to match the vector store dimension —
        the caller (KnowledgeBase) is responsible for NOT writing these
        hash vectors into the semantic store when no real embedding key exists.
        """
        np.random.seed(int(hashlib.md5(text.encode()).hexdigest(), 16) % (2**31))
        return np.random.randn(self._dim).tolist()

File: internal/rag/embedder/embedder.go
// Package embedder mirrors diagflow/rag/embedder.py — text → embedding vectors.
package embedder

import (
	"crypto/md5"
	"math/rand"
)

// Embedder produces embeddings, falling back to a deterministic hash vector
// when no API key is present.
type Embedder struct {
	Model  string
	APIKey string
	dim    int
}

// New builds an Embedder. embeddingDim must match the vector store dimension.
func New(model, apiKey string, embeddingDim int) *Embedder {
	return &Embedder{Model: model, APIKey: apiKey, dim: embeddingDim}
}

// Embed produces a vector. With an API key it calls OpenAI; otherwise it uses
// the deterministic fallback.
func (e *Embedder) Embed(text string) []float64 {
	if e.APIKey != "" {
		return e.openaiEmbed(text)
	}
	return e.fallbackEmbed(text)
}

// EmbedBatch embeds multiple texts.
func (e *Embedder) EmbedBatch(texts []string) [][]float64 {
	out := make([][]float64, len(texts))
	for i, t := range texts {
		out[i] = e.Embed(t)
	}
	return out
}

// HasRealEmbeddings reports whether we can produce semantically meaningful
// vectors. DEGRADED MODE: the OpenAI embedding call is not yet wired (see
// openaiEmbed), so this is always false — the KB semantic fast-path in
// DiagAgent Phase 1 stays disabled and only the deterministic fallback is
// used. Flip this to true ONLY after wiring a real embedding provider.
func (e *Embedder) HasRealEmbeddings() bool {
	return false
}

// openaiEmbed is a stub returning the deterministic fallback. TODO: wire a
// real embedding provider (POST /v1/embeddings) before enabling
// HasRealEmbeddings.
func (e *Embedder) openaiEmbed(text string) []float64 {
	return e.fallbackEmbed(text)
}

// fallbackEmbed is a deterministic hash-based vector matching the store dim.
// Not semantically meaningful — the caller must not write these into a real
// vector store when no embedding key is present.
func (e *Embedder) fallbackEmbed(text string) []float64 {
	sum := md5.Sum([]byte(text))
	seedBytes := sum[:]
	r := rand.New(rand.NewSource(int64FromBytes(seedBytes)))
	out := make([]float64, e.dim)
	for i := range out {
		out[i] = r.NormFloat64()
	}
	return out
}

func int64FromBytes(b []byte) int64 {
	var v int64
	for i, c := range b[:4] {
		v |= int64(c) << (8 * i)
	}
	return v
}
